streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1848,113 +1907,101 @@ object ZChannel {
-                               ZIO.scopedWith { scope =>
-                                 (queueReader >>> channel)
-                                   .toPullIn(scope)
-                                   .flatMap(
-                                     evaluatePull(_)
-                                       .raceAwait(errorSignal.await.interruptible)
-                                       .raceAwait(canceler.await.interruptible)
-                                   )
-                               }
-                             childFiber <- permits
-                                             .withPermit(latch.succeed(()) *> raceIOs)
-                                             .forkIn(scope)
-                             _       <- latch.await
-                             errored <- errorSignal.isDone
-                           } yield !errored
+                           pull.flatMap(channel =>
+                             for {
+                               canceler <- Promise.make[Nothing, Unit]
+                               latch    <- Promise.make[Nothing, Unit]
+                               size     <- cancelers.size
+                               _        <- ZIO.when(size >= n)(cancelers.take.flatMap(_.succeed(())))
+                               _        <- cancelers.offer(canceler)
+                               raceIOs =
+                                 ZIO.scopedWith { scope =>
+                                   (incoming >>> channel)
+                                     .toPullInAlt(scope)
+                                     .flatMap(evaluatePull(_).race(canceler.await))
+                                 }
+                               childFiber <- permits
+                                               .withPermit(latch.succeed(()) *> raceIOs)
+                                               .fork
+                               _ <- latch.await
+                             } yield ()
+                           )
                        }
-                   }
-                 )
-                 .repeatWhileEquals(true)
-                 .forkIn(scope)
-        } yield (queue, input)
-      }.map { case (queue, input) =>
+        _ <- pullStrategy.forever
+               .catchAllCause(cause =>
+                 cause.failureOrCause match {
+                   case Left(_: Left[OutErr, OutDone]) => outgoing.offer(Exit.failCause(cause))
+                   case Left(x: Right[OutErr, OutDone]) =>
+                     permits.withPermits(n.toLong)(ZIO.unit) *>
+                       lastDone.get.flatMap {
+                         case Some(lastDone) => outgoing.offer(Exit.fail(Right(f(lastDone, x.value))))
+                         case None           => outgoing.offer(Exit.fail(Right(x.value)))
+                       }
+                   case Right(cause) => outgoing.offer(Exit.failCause(cause.map(Left(_))))
+                 }
+               )
+               .race(errorSignal.await)
```

Comment:
See my other comment, we should change this to `errorSignal.await.interruptible`

Reply:
This I struggle with that a lot. It means that every time we use `await` in an API, we must not forget to add `interruptible` to it, otherwise someone might call the API within an uninterruptible region and completely change the semantics.